src/Datadog.Trace.ClrProfiler.Managed/Integrations/ElasticsearchNet5Integration.cs

```diff
@@ -0,0 +1,124 @@
+        /// <returns>The original result</returns>
+        [InterceptMethod(
+            CallerAssembly = "Elasticsearch.Net",
+            TargetAssembly = "Elasticsearch.Net",
+            TargetType = "Elasticsearch.Net.IRequestPipeline",
+            TargetMinimumVersion = TargetVersion,
+            TargetMaximumVersion = TargetVersion)]
+        public static object CallElasticsearch<TResponse>(object pipeline, object requestData)
+        {
+            // ElasticsearchResponse<TReturn> CallElasticsearch<TReturn>(RequestData requestData) where TReturn : class;
+            var originalMethod = Emit.DynamicMethodBuilder<Func<object, object, TResponse>>
+                                     .GetOrCreateMethodCallDelegate(
+                                          pipeline.GetType(),
+                                          "CallElasticsearch",
+                                          methodGenericArguments: new[] { typeof(TResponse) });
+
+            using (var scope = ElasticsearchNetCommon.CreateScope(Tracer.Instance, IntegrationName, pipeline, requestData))
+            {
+                try
+                {
+                    return originalMethod(pipeline, requestData);
+                }
+                catch (Exception ex) when (scope?.Span.SetExceptionForFilter(ex) ?? false)
+                {
+                    // unreachable code
+                    throw;
+                }
+            }
+        }
+
+        /// <summary>
+        /// Traces an asynchronous call to Elasticsearch.
+        /// </summary>
+        /// <typeparam name="TResponse">Type type of the response</typeparam>
+        /// <param name="pipeline">The pipeline for the original method</param>
+        /// <param name="requestData">The request data</param>
+        /// <param name="cancellationTokenSource">A cancellation token</param>
+        /// <returns>The original result</returns>
+        [InterceptMethod(
+            CallerAssembly = "Elasticsearch.Net",
+            TargetAssembly = "Elasticsearch.Net",
+            TargetType = "Elasticsearch.Net.IRequestPipeline",
+            TargetMinimumVersion = TargetVersion,
+            TargetMaximumVersion = TargetVersion)]
+        public static object CallElasticsearchAsync<TResponse>(object pipeline, object requestData, object cancellationTokenSource)
+        {
+            // Task<ElasticsearchResponse<TReturn>> CallElasticsearchAsync<TReturn>(RequestData requestData, CancellationToken cancellationToken) where TReturn : class;
+            var tokenSource = cancellationTokenSource as CancellationTokenSource;
+            var cancellationToken = tokenSource?.Token ?? CancellationToken.None;
+
+            // return CallElasticsearchAsyncInternal<TResponse>(pipeline, requestData, cancellationToken);
+
+            var pipelineType = pipeline.GetType();
+            var responseType = pipelineType.Assembly.GetType("Elasticsearch.Net.ElasticsearchResponse`1", throwOnError: false);
```

Comment:
Are there any instances of anything other than this response type being used in elasticsearch.net or nest?
The restriction is on `IElasticsearchResponse`, so the incoming method may not have this type.

https://github.com/elastic/elasticsearch-net/blob/67fe8ce6cbaea8ab593dfe77d038fa5b6bb4f903/src/Elasticsearch.Net/Transport/Pipeline/IRequestPipeline.cs#L25

Reply:
Good catch. I'll look into this.

Reply:
In [the 5.x branch](https://github.com/elastic/elasticsearch-net/blob/5.x/src/Elasticsearch.Net/Transport/Pipeline/IRequestPipeline.cs), the return type is `Task<ElasticsearchResponse<TReturn>>` and `ElasticsearchResponse` (not `TReturn`/`TResponse`) implements `IElasticsearchResponse`, so I think this is fine (for 5.x only).